test/hotspot/jtreg/compiler/integerArithmetic/TestNegMultiply.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2021, Red Hat, Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @bug 8273454
+ * @summary Test transformation (-a)*(-b) = a*b
+ *
+ * @run main/othervm -XX:-TieredCompilation -XX:-BackgroundCompilation -XX:-UseOnStackReplacement TestNegMultiply
+ *
+ */
+
+import java.util.Random;
+
+public class TestNegMultiply {
+    private static Random random = new Random();
+    private static final int TEST_COUNT = 2000;
+
+    private static int test(int a, int b) {
+        return (-a) * (-b);
+    }
+
+    private static void testInt(int a, int b) {
+        int expected = (-a) * (-b);
+        for (int i = 0; i < 20_000; i++) {
```

Comment:
Why do you need a second loop in here? It's sufficient to set `TEST_COUNT` high enough to trigger compilation. I would suggest something like this:

```
private static int testInt(int a, int b) {
    return (-a) * (-b);
}

private static void runIntTests() {
    for (int i = 0; i < TEST_COUNT; i++) {
        int a = random.nextInt();
        int b = random.nextInt();
        int res = testInt(a, b);
        Asserts.assertEQ(a * b, res);
    }
}
```

And then run with `-XX:CompileCommand=dontinline,TestNegMultiply::test*`. No need to disable OnStackReplacement.

Reply:
The inner loop ensures that all tests hit JIT-ed version. If the transformation is broken, I would prefer the test fails for the very first iteration, instead of somewhere in the middle.

I refactored the code to remove inner loop.

Reply:
You can't control the iteration in which the test would fail if there's a bug in C2 (it could only fail for **some** random values). Therefore, you could as well use random values for the warmup and simply increase `TEST_COUNT` to ensure that C2 compilation is triggered and we run a reasonable amount of iterations with C2 compiled code. 

Your newest version of the test now has the problem that OSR compilation might C2 compile the computation of the expected value and then you are comparing the output of a C2 compiled method to a C2 compiled method instead of the interpreter. You have the following options:
- Compute the expected value as `a * b`. In that case it's fine if the computation is C2 compiled as well.
- Prevent compilation of the `run*` methods (either by disabling OSR compilation or by completely disabling compilation of these methods)

Reply:
And sorry for being picky here but I would like to keep tests as simple as possible :)

Reply:
Fixed according to you comments.

I really appreciate you suggestions, thanks!